### PR TITLE
Add Oracle Coherence embedding store implementation

### LINKED ISSUE #2002: Add embedding store implementation for Oracle Coherence
**Describe the solution you'd like**

We added support for vector storage and search in [Oracle Coherence CE 24.09](https://github.com/oracle/coherence), and would like to contribute embedding store implementation. Opening this issue so we can submit a PR.

### PR DESCRIPTION
**Pull request by aseovic**
## Issue

Closes #2002 

## Change
Added `langchain4j-coherence` module that provides embedding store and chat memory store implementations that use [Oracle Coherence Community Edition](https://github.com/oracle/coherence) as a store.

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [*] I have manually run all the unit and integration tests in the [core] (https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green (they are not, but I doubt it had anything to do with my changes -- Jacoco is failing the build because `core` requires 75% code coverage and it's only 68% at the moment)

- [x] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

We have all three of the above ready to submit, but will wait until this PR is reviewed and approved before submitting them.

## Checklist for adding new maven module

- [X] I have added my new module in the root `pom.xml` and `langchain4j-bom/pom.xml`

## Checklist for adding new embedding store integration

- [X] I have added a `CoherenceEmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [X] I have added a `CoherenceEmbeddingStoreRemovalIT` that extends from `EmbeddingStoreWithRemovalIT`

### CODE REVIEW

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+        String text = chunk.text();
+        TextSegment segment = text == null ? null : new TextSegment(chunk.text(), getMetadata(chunk.metadata()));
+        Vector<float[]> vector = chunk.vector();
+        Embedding embedding = vector == null ? null : new Embedding(vector.get());
+
+        return new EmbeddingMatch<>(score, key, embedding, segment);
+    }
+
+    private static Metadata getMetadata(Map<String, Object> mapMetadata) {
+        Metadata metadata = new Metadata();
```
Comment (Martin7-1) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]:
```suggestion
        mapMetadata.entrySet().removeIf(entry -> entry.getValue() == null);
        return Metadata.from(mapMetadata);
```

`Metadata#add` is deprecated, so can we just filter and construct it using `Metadata#from`?

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
+package dev.langchain4j.store.memory.chat.coherence;
+
+import com.oracle.coherence.ai.DocumentChunk;
+import com.oracle.coherence.ai.Vector;
```
Comment (Martin7-1) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java]: Looks like there are some unused imports, can we optimzie it?

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+     *
+     * @param filter  the LangChain filter to convert
+     *
+     * @return the equivalent Coherence filter
+     */
+    static com.tangosol.util.Filter<DocumentChunk> map(Filter filter) {
+        if (filter == null) {
+            return null;
+        } else if (filter instanceof IsEqualTo) {
+            return mapEqual((IsEqualTo) filter);
```
Comment (Martin7-1) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java]: minor: as we already migrated to JDK17, we can use [pattern matching for instanceof](https://openjdk.org/jeps/394) now.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import dev.langchain4j.store.embedding.filter.comparison.IsNotIn;
+import dev.langchain4j.store.embedding.filter.logical.And;
+import dev.langchain4j.store.embedding.filter.logical.Not;
+import dev.langchain4j.store.embedding.filter.logical.Or;
+
+import java.util.Collection;
+import java.util.HashSet;
+
+/**
+ * A mapper that maps {@link Filter LangChain filters} to
```
Comment (Martin7-1) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java]:
```suggestion
 * A mapper that maps {@link Filter LangChain4j filters} to
```

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ * {@link com.tangosol.util.Filter Coherence filters} that
+ * can be applied to {@link DocumentChunk} metadata.
+ */
+class CoherenceMetadataFilterMapper
+    {
+    /**
+     * Return the Coherence filter that is equivalent to the
+     * specified LangChain filter.
```
Comment (Martin7-1) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceMetadataFilterMapper.java]:
```suggestion
     * specified LangChain4j filter.
```

File: pom.xml [hunk 6]
```diff
@@ -29,6 +29,7 @@
         <module>langchain4j-azure-open-ai</module>
         <module>langchain4j-bedrock</module>
         <module>langchain4j-chatglm</module>
+        <module>langchain4j-coherence</module>
```
Comment (Martin7-1) [inline pom.xml]: minor: please move it to `<!-- embedding stores -->` block.

**Review by Martin7-1 — COMMENTED** [review]
Thank you! And we should consider the lifecycle of `session`, looks like it's not closed properly now.

**Comment by aseovic** [comment]
@Martin7-1 Which `session` lifecycle are you talking about? If it's Coherence `Session`, there really isn't anything to "close" -- it's more of a virtual than physical construct.

File: docs/docs/integrations/embedding-stores/index.md [hunk 7]
```diff
@@ -13,6 +13,7 @@ sidebar_position: 0
 | [Azure CosmosDB NoSQL](/integrations/embedding-stores/azure-cosmos-nosql)             | ✅                |                            |                     |
 | [Cassandra](/integrations/embedding-stores/cassandra)                                 | ✅                |                            |                     |
 | [Chroma](/integrations/embedding-stores/chroma)                                       | ✅                | ✅                          | ✅                   |
+| [Coherence](/integrations/embedding-stores/coherence)                          | ✅                | ✅                     | ✅                   |
```
Comment (langchain4j) [inline docs/docs/integrations/embedding-stores/index.md:16]:
```suggestion
| [Oracle Coherence](/integrations/embedding-stores/oracle-coherence)                          | ✅                | ✅                     | ✅                   |
```

Comment (langchain4j) [reply docs/docs/integrations/embedding-stores/index.md:16]: What about naming it Oracle Coherence?
This will require moving it below close to Oracle

File: docs/docs/integrations/embedding-stores/oracle-coherence.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+https://coherence.community/
+
+
+## Maven Dependency
+
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-coherence</artifactId>
```
Comment (langchain4j) [inline docs/docs/integrations/embedding-stores/oracle-coherence.md]:
```suggestion
    <artifactId>langchain4j-oracle-coherence</artifactId>
```

File: docs/docs/integrations/embedding-stores/oracle-coherence.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
+---
+sidebar_position: 7
```
Comment (langchain4j) [inline docs/docs/integrations/embedding-stores/oracle-coherence.md]: nit: position should probably be updated if it is moved below

File: docs/docs/integrations/embedding-stores/oracle-coherence.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+            <scope>import</scope>
+        </dependency>
+    </dependencies>
+</dependencyManagement>
+
+<dependencies>
+    <dependency>
+        <groupId>dev.langchain4j</groupId>
+        <artifactId>langchain4j-coherence</artifactId>
+        <version>0.32.0</version>
```
Comment (langchain4j) [inline docs/docs/integrations/embedding-stores/oracle-coherence.md]:
```suggestion
        <version>0.36.0</version>
```

File: langchain4j-coherence/pom.xml [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+        <dependency>
+            <groupId>dev.langchain4j</groupId>
+            <artifactId>langchain4j-embeddings-all-minilm-l6-v2-q</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>${coherence.groupId}</groupId>
+            <artifactId>coherence</artifactId>
```
Comment (langchain4j) [inline langchain4j-coherence/pom.xml]: nit: please move this and coherence-hnsw above test dependencies

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 12]
```diff
@@ -0,0 +1,526 @@
...
+
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.Utils.isNullOrEmpty;
+import static dev.langchain4j.internal.ValidationUtils.ensureTrue;
+
+/**
+ * An {@link EmbeddingStore} backed by an Oracle Coherence {@link NamedMap}.
+ */
+public class CoherenceEmbeddingStore
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java:54]:
```suggestion
public class OracleCoherenceEmbeddingStore
```

File: langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreIT.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+import dev.langchain4j.store.embedding.EmbeddingStoreIT;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+/**
+ * An extension of the LangChain4J {@link EmbeddingStoreIT} tests
+ * that use Coherence as an {@link EmbeddingStore}.
+ */
+public class CoherenceEmbeddingStoreIT
+        extends EmbeddingStoreIT {
```
Comment (langchain4j) [inline langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreIT.java]:
```suggestion
        extends EmbeddingStoreWithFilteringIT {
```
This test class can inherit from `EmbeddingStoreWithFilteringIT` and get all the required tests

File: langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreWithFilteringIT.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import dev.langchain4j.store.embedding.EmbeddingStoreWithFilteringIT;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+/**
+ * An extension of the LangChain4J {@link EmbeddingStoreWithFilteringIT} tests
+ * that use Coherence as an {@link EmbeddingStore}.
+ */
+public class CoherenceEmbeddingStoreWithFilteringIT
```
Comment (langchain4j) [inline langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreWithFilteringIT.java]: This test class can be removed as there is `CoherenceEmbeddingStoreIT` that should inherit all required tests

File: langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreWithoutMetadataIT.java [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+import dev.langchain4j.store.embedding.EmbeddingStore;
+import dev.langchain4j.store.embedding.EmbeddingStoreWithoutMetadataIT;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+/**
+ * An extension of the LangChain4J {@link EmbeddingStoreWithoutMetadataIT} tests
+ * that use Coherence as an {@link EmbeddingStore}.
+ */
+public class CoherenceEmbeddingStoreWithoutMetadataIT
```
Comment (langchain4j) [inline langchain4j-coherence/src/test/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStoreWithoutMetadataIT.java]: This test class can be removed as there is `CoherenceEmbeddingStoreIT` that should inherit all required tests

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+/**
+ * A {@link ChatMemoryStore} backed by an Oracle Coherence named map.
+ * <p>
+ * The {@link CoherenceChatMemoryStore} supports memory identifiers of any type
+ * that is a valid Coherence {@link NamedMap} key. The key type must properly
+ * implement {@code equals()} and {@code hashCode()} and be serializable by
+ * the configured Coherence serializer.
+ */
+public class CoherenceChatMemoryStore
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java:26]:
```suggestion
public class OracleCoherenceChatMemoryStore
```

File: langchain4j-coherence/src/test/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStoreIT.java [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        // given
+        List<ChatMessage> messages = memoryStore.getMessages(userId);
+        assertThat(messages).isEmpty();
+
+        // when
+        List<ChatMessage> chatMessages = new ArrayList<>();
+        chatMessages.add(new SystemMessage("You are a large language model working with Langchain4j"));
+        List<Content> userMsgContents = new ArrayList<>();
+        userMsgContents.add(new ImageContent("someCatImageUrl"));
+        chatMessages.add(new UserMessage("What do you see in this image?", userMsgContents));
```
Comment (langchain4j) [inline langchain4j-coherence/src/test/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStoreIT.java]: nit: "What do you see in this image?" is a name of the user, not the text content

File: langchain4j-coherence/src/test/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStoreIT.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        memoryStore.updateMessages(userId, chatMessages);
+
+        // then
+        messages = memoryStore.getMessages(userId);
+        assertThat(messages).hasSize(2);
```
Comment (langchain4j) [inline langchain4j-coherence/src/test/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStoreIT.java:89]: Could you please add a bit more assertions? E.g. that returned messages are identical to persisted ones

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 19]
```diff
@@ -0,0 +1,526 @@
...
+
+    /**
+     * Create a {@link Comparator} that sorts a double in reverse order.
+     *
+     * @param keyExtractor  the function to use to extract a double value
+     * @param <T>           the type of value the function extracts from
+     *
+     * @return the result of comparing two extracted doubles in reverse order
+     */
+    public static<T> Comparator<T> reverseComparingDouble(ToDoubleFunction<? super T> keyExtractor) {
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]:
```suggestion
    private static<T> Comparator<T> reverseComparingDouble(ToDoubleFunction<? super T> keyExtractor) {
```

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 20]
```diff
@@ -0,0 +1,526 @@
...
+
+        /**
+         * Set whether to force normalization of vectors on adding and searching.
+         *
+         * @param f  {@code true} if the {@link CoherenceEmbeddingStore} should call force
+         *           {@link Embedding#normalize()} on embeddings when adding or searching
+         *
+         * @return this builder for fluent method calls
+         */
+        public Builder forceNormalize(boolean f) {
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]:
```suggestion
        public Builder normalizeEmbeddings(boolean f) {
```

nit: can we rename it? There is already `VearchEmbeddingStore` that uses `normalizeEmbeddings` name

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 21]
```diff
@@ -0,0 +1,526 @@
...
+        for (QueryResult<DocumentChunk.Id, DocumentChunk> result : results) {
+            double score = RelevanceScore.fromCosineSimilarity(1.0f - result.getDistance());
+            if (checkMin && score < minScore) {
+                continue;
+            }
+
+            DocumentChunk.Id id = result.getKey();
+            DocumentChunk chunk = result.getValue();
+
+            boolean fHighlyRelevant = score > 0.75f;
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: where does this value comes from? Why is treatment different in this case?

**Review by ghost — COMMENTED** [review]
@aseovic thank you!

File: docs/docs/integrations/embedding-stores/index.md (hunk 7 above)
Comment (aseovic) [reply docs/docs/integrations/embedding-stores/index.md:16]: I'd prefer not to -- it's too long, and all of our artifacts are just `coherence*`, not `oracle-coherence`, so leaving it the way it is now feels more natural.

File: docs/docs/integrations/embedding-stores/oracle-coherence.md (hunk 8 above)
Comment (aseovic) [reply docs/docs/integrations/embedding-stores/oracle-coherence.md]: Same as above.

File: docs/docs/integrations/embedding-stores/oracle-coherence.md (hunk 10 above)
Comment (aseovic) [reply docs/docs/integrations/embedding-stores/oracle-coherence.md]: Fixed the version and renamed the file to `coherence.md`, for consistency (and because that's what `index.md` links to ;-)

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 12 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java:54]: Same as above -- I'd rather not.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java (hunk 16 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/memory/chat/coherence/CoherenceChatMemoryStore.java:26]: Rather not...

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 21 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: Tbh, I have no idea... 

@thegridman Thoughts?

Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: Ah, actually, I do -- this is something I was playing with in our RAG implementation, but I don't think it's necessary (or desired) here. I'll remove it, good catch.

**Comment by aseovic** [comment]
@langchain4j Thanks for the quick turnaround and a thorough review.

The only remaining question is whether to use Coherence or Oracle Coherence, and apply changes you suggested above if we choose the latter. 

I prefer to leave it as is, but if you have strong feelings about it I have no problem renaming it either, so let me know what you want us to do.

**Comment by aseovic** [comment]
I'm also not sure what you want me to do with Codacy issues.

I've addressed the doc ones, but the remaining ones are related to test naming. We followed the existing examples and code style, but Codacy rules seem to be different. Let me know what you want us to do.

File: docs/docs/integrations/embedding-stores/coherence.md [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+
+<https://coherence.community/>
+
+## Maven Dependency
+
+```xml
+<dependency>
+    <groupId>dev.langchain4j</groupId>
+    <artifactId>langchain4j-coherence</artifactId>
+    <version>0.32.0</version>
```
Comment (langchain4j) [inline docs/docs/integrations/embedding-stores/coherence.md]:
```suggestion
    <version>0.36.0</version>
```

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 23]
```diff
@@ -0,0 +1,506 @@
...
+            if (checkMin && score < minScore) {
+                continue;
+            }
+
+            matches.add(embeddingMatch(score, result.getKey(), result.getValue()));
+        }
+
+        matches.sort(COMPARATOR);
+        return new EmbeddingSearchResult<>(matches.size() > request.maxResults()
+                                           ? matches.subList(0, request.maxResults())
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: Do I understand correctly that `results` will contain **_all_** the chunks that were ever added? And filtering by `minScore` and `maxResults` happens here? Isn't it very inefficient?

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 24]
```diff
@@ -0,0 +1,506 @@
...
+
+    /**
+     * Create an {@link CoherenceEmbeddingStore}.
+     * <p>
+     * This method is protected, instances of {@link CoherenceEmbeddingStore}
+     * are created using the builder.
+     *
+     * @param namedMap             the {@link NamedMap} to contain the {@link DocumentChunk document chunks}
+     * @param normalizeEmbeddings  {@code true} if this {@link CoherenceEmbeddingStore} should call
+     *                             {@link Embedding#normalize()} on embeddings when adding or searching
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java:81]:
```suggestion
     *                             {@link Embedding#normalize()} on embeddings when adding them to the store
```
It does not seeem that it happens when searching, also it does not seem to be required on search.

Please fix it everywhere in this class (5 times)

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 25]
```diff
@@ -0,0 +1,506 @@
...
+
+    /**
+     * Perform a KNN search.
+     *
+     * @param vector  the vector to find the nearest neighbours to
+     * @param k       the maximum number of neighbours to find
+     *
+     * @return  the search results
+     */
+    public List<QueryResult<DocumentChunk.Id, DocumentChunk>> search(float[] vector, int k) {
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: Why is this needed?

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 26]
```diff
@@ -0,0 +1,506 @@
...
+         return documentChunks.aggregate(new SimilaritySearch<>(DocumentChunk::vector, new Float32Vector(vector), k));
+     }
+
+    /**
+     * Obtain the {@link NamedMap} used to store embeddings.
+     *
+     * @return the {@link NamedMap} used to store embeddings
+     */
+    public NamedMap<DocumentChunk.Id, DocumentChunk> getDocumentChunks() {
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]:
```suggestion
    public NamedMap<DocumentChunk.Id, DocumentChunk> documentChunks() {
```

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java [hunk 27]
```diff
@@ -0,0 +1,506 @@
...
+        String key = id.toString();
+        String text = chunk.text();
+        TextSegment segment = text == null ? null : new TextSegment(chunk.text(), getMetadata(chunk.metadata()));
+        Vector<float[]> vector = chunk.vector();
+        Embedding embedding = vector == null ? null : new Embedding(vector.get());
+
+        return new EmbeddingMatch<>(score, key, embedding, segment);
+    }
+
+    private static Metadata getMetadata(Map<String, Object> mapMetadata) {
```
Comment (langchain4j) [inline langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: nit: could have a better name

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 23 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: No, we pass `request.maxResults()` as an argument to `SimilaritySearch` aggregator, which is what performs the actual search across the cluster in parallel, and returns only the specified number of best matches.

I believe the above was to handle the case that I removed during last review cycle, which was adding previous and next chunk as well for "highly relevant" results, which would've increased the total number of results and pushed some out, so we had to trim the list of matches at the end.

I don't think that is necessary any longer, and I don't think we need to sort the matches either, as they should already be sorted by the `SimilaritySearch` aggregator. I'll simplify/remove the code above, good catch.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 24 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java:81]: We actually do normalize query embedding if this field is set (see L161-163), and I do think it's required to ensure that we are comparing apples to apples (ie. if the embeddings in the store are normalized, we need to normalize search vector/embedding as well).

So the doc above is correct, unless we change the behavior.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 25 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: It's not. Removed.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 26 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: This is not needed either. Removed altogether.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 27 above)
Comment (aseovic) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java]: Agreed. Changed it to `mapToMetadata` and added javadoc.

File: langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java (hunk 24 above)
Comment (langchain4j) [reply langchain4j-coherence/src/main/java/dev/langchain4j/store/embedding/coherence/CoherenceEmbeddingStore.java:81]: Sorry, I misunderstood it

**Comment by aseovic** [comment]
@langchain4j Anything else preventing this from moving out of draft and getting merged?

I also have an example and Spring Boot starter to submit PRs for. Not sure when you want me to submit those?

**Comment by langchain4j** [comment]
@aseovic I'll try to review the latest version today and let you know, but I think it should be good to go

**Review by ghost — APPROVED** [review]
@aseovic thank you!

**Comment by langchain4j** [comment]
Hi @aseovic, I've noticed that `com.oracle.coherence.ce:coherence` has a "Universal Permissive License v 1.0" license.
Which license should `langchain4j-coherence` module have? Thanks!

**Comment by aseovic** [comment]
@langchain4j Yeah, our own projects use UPL 1.0 (Oracle mandate), but contributions to other projects typically use that project's license, so in this case Apache 2.0 (which is compatible with UPL 1.0).

Normally, we would add Apache 2.0 license headers to the source files to make it clear, but we didn't see any in other code, including Oracle DB vector store integration, so we omitted them. If there is any place you want us to add this information to, let me know and I'll do it.
